### ISSUE TITLE
very low accuracy test on one GPU

### ISSUE THREAD

**Issue by zhj123169** [leoxiaobin/deep-high-resolution-net.pytorch#25]
I don't have four GPUs.If I run the command that you provide:
```
python tools/test.py \
    --cfg experiments/coco/hrnet/w32_256x192_adam_lr1e-3.yaml \
    TEST.MODEL_FILE models/pytorch/pose_coco/pose_hrnet_w32_256x192.pth \
    TEST.USE_GT_BBOX False
```
i got an error about device_ids.
So I change the code in test.py: 
model = torch.nn.DataParallel(model,device_ids=cfg.GPUS).cuda()        to
model = torch.nn.DataParallel(model, device_ids=(0,)).cuda()

the other:
num_workers=cfg.WORKERS        to
 num_workers=0,

It works but the model 'pose_hrnet_w32_256x192.pth' gets very low accuracy(nearly 0).I think somewhere I changed wrong or some code still need to be changed.Could you help me?

**Comment by YinRui1991** [comment]
@leoxiaobin Thanks for the excellenct codes !
@zhj123169 hi I run the test.py just now, and I get the following results.
 
Test: [0/814]	Time 7.468 (7.468)	Loss 0.0002 (0.0002)	Accuracy 0.000 (0.000)
Test: [100/814]	Time 0.720 (0.829)	Loss 0.0002 (0.0001)	Accuracy 0.000 (0.003)
Test: [200/814]	Time 0.722 (0.790)	Loss 0.0001 (0.0001)	Accuracy 0.000 (0.003)
Test: [300/814]	Time 0.723 (0.783)	Loss 0.0000 (0.0001)	Accuracy 0.000 (0.002)
Test: [400/814]	Time 0.737 (0.778)	Loss 0.0001 (0.0001)	Accuracy 0.000 (0.003)
Test: [500/814]	Time 0.724 (0.776)	Loss 0.0000 (0.0001)	Accuracy 0.000 (0.003)
Test: [600/814]	Time 0.779 (0.773)	Loss 0.0001 (0.0001)	Accuracy 0.000 (0.003)
Test: [700/814]	Time 0.726 (0.772)	Loss 0.0005 (0.0001)	Accuracy 0.000 (0.003)
Test: [800/814]	Time 0.721 (0.770)	Loss 0.0002 (0.0001)	Accuracy 0.059 (0.003)
=> writing results json to output/coco/pose_hrnet/w32_256x192_adam_lr1e-3/results/keypoints_val2017_results_0.json
Loading and preparing results...
DONE (t=4.88s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=10.66s).
Accumulating evaluation results...
DONE (t=0.39s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.744
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.905
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.819
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.708
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.810
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.798
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.942
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.865
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.757
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.858
| Arch | AP | Ap .5 | AP .75 | AP (M) | AP (L) | AR | AR .5 | AR .75 | AR (M) | AR (L) |
|---|---|---|---|---|---|---|---|---|---|---|
| pose_hrnet | 0.744 | 0.905 | 0.819 | 0.708 | 0.810 | 0.798 | 0.942 | 0.865 | 0.757 | 0.858 |

I don't know the accuracy represents what, but the Ap and Ar are the same with that are reported in README. I will check the code to find what accuracy represents.

**Comment by zhj123169** [comment]
@YinRui1991  The results I got are very similar to yours,but Ap and Ar it shows is  person detection result of COCO val2017. It is not the results of running on our GPU.Something still runs wrong.If you have solved it,could you share it with me?Appreciate it!

**Comment by YinRui1991** [comment]
@zhj123169 I think the accuracy represents person detection performance, and for the top-down method it is necessary to have a person detector. The command set "TEST.USE_GT_BBOX False" means the person bboxes are offered by "COCO_val2017_detections_AP_H_56_person.json". So I think you don't care about the accuracy and loss. All the top-down methods have this situation (such as CPN). If you want to use HRNET, you need have a good person detector. In my progects, I use YOLOV3 for object detection. I plan to test speed and precision of HRNET(CPN is too slow, and I see HRNET can reach 100 fps for 2 two-person images at issues26 (https://github.com/leoxiaobin/deep-high-resolution-net.pytorch/issues/26).

**Comment by zhj123169** [comment]
@YinRui1991   Thank you for your help.I haven't tried it with detector yet.Someone else also suggests me using YOLO detector.I will try it  in the next stage!Thank you!

**Comment by JayLSU** [comment]
@YinRui1991 Have you implemented YOLO detector with HRnet? I am going to do this, since I wish to implement a real time pose estimation application.  Could you provide some advice? Thanks.

**Comment by YinRui1991** [comment]
@JayLSU Yes, I have realized YOLO+HRnet, you can read this https://github.com/Microsoft/human-pose-estimation.pytorch/issues/26 for reference (baseline keypoints demo). But the HRnet performs not well on my own testing videos. I find it is sensitive to the person bbox. I mean you give two different but similar bboxes of the same person, the key-points locations results may different. Maybe there are some person crop preprocessing work needs to confirm (get person crop from whole image, and padding).

**Comment by JayLSU** [comment]
@YinRui1991 Thanks, I will take a look at that reference. For the issue you mentioned, it is quite like what I experienced in reproducing the [LSTM-pose-machine](https://github.com/lawy623/LSTM_Pose_Machines). Before using it for pose estimation, the person has to be cropped out, resize, and so on. You can take a look at that. I am very interested in this area and would like to have some peer to discuss, may I have your email for later discussion or you can email me (sguo_autonomous@163.com) if you think that is more comfortable way? Thanks again. Bests.

**Comment by YinRui1991** [comment]
@JayLSU Hi, this is my email (yinrui_cf@163.com), I'm glad to communicate pose estimation with you.

**Comment by aliYaHf** [comment]
@YinRui1991  Hi, have you  trained on MPII dataset? I got a very low AR

**Comment by YinRui1991** [comment]
@aliYaHf Sorry, I don't try to train HRnet with MPII dataset. Do you try the model offered by leoxiaobin?

**Comment by aliYaHf** [comment]
@YinRui1991 Thank you for your reply, I tested the trained model by leoxiaobiao, and the results are very close to those in README. I reduced batch_size to 16 and used MPII data provided by leo, but the results of the training varied greatly.

**Comment by YinRui1991** [comment]
@aliYaHf If you want to train a model that has the similar performance as the author's, I think you need to find some information in the paper, such as hyper-parameters, batch-size, gpu numbers, iterations... And if finetune with model trained by other dataset.
Or you can finetune with the offered MPII model to see the loss value, and compared with yours. 
For your unstable loss value, you can decrease the learning rate and use bigger batch-size(if your graphics card enough memory, or calculate loss with multi-batches iterations) to let the loss value not oscillate.

**Comment by aliYaHf** [comment]
@YinRui1991 Thank you. Because I don't have enough memory on my graphics card, so i reduced the batch_size, I‘m not sure if that's the reason result in the model doesn't converge. I'll try what you mentioned, thank you again.

**Comment by Muhtasham** [comment]
> @YinRui1991 Thank you. Because I don't have enough memory on my graphics card, so i reduced the batch_size, I‘m not sure if that's the reason result in the model doesn't converge. I'll try what you mentioned, thank you again.

Hello, can you please tell how to reduce batch size? Which line to change in the code?
